How many late days are you using for this assignment? 

In [48]:
# Imports and pip installations (if needed)
import pandas as pd
import sklearn as skl
from sklearn.preprocessing import OneHotEncoder

# Part 1: Load the dataset

In [2]:
# Load the given datasets
data1 = pd.read_csv("https://drive.google.com/uc?export=download&id=1BotQgclZm7Anu4QcGKkDZblvoxIpSlbW")
data2 = pd.read_csv("https://drive.google.com/uc?export=download&id=1De2YdaNC07zPqkZZjm0Po-40vPYikZFd")

# Print the data
print(data1.head())
print(data2.head())
print(data1.info())
print(data2.info())

   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    normal    normal  notpresent  notpresent   no   no   

  cad appet   pe  ane  Target  
0  no  good  yes   no     ckd  
1  no  good   no   no  notckd  
2  no  good  yes   no     ckd  
3  no  good   no  yes     ckd  
4  no  good   no   no  notckd  
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     76411

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
In this data set we have 24 features all a mix of binary and categorical and the target variable - class - predicting whether a person has chronic kidney disease. The features are all health-related information like blood urea level, sodium level, whether the person is anemic or not, etc. The features are either binary (0 or 1), integer, continuous or categorical. Some numerical features are age, blood pressure, red blood cells, etc. The categorical features include specific gravity, albumin, and sugar.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
print(data1.dtypes)
print(data2.dtypes)
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
data2_duplicates = data2.duplicated().any()
print("Are there any duplicate rows in data2? ", data2_duplicates)
print("What are the total number of duplicate rows in data2? ", data2.duplicated().sum())

# Drop these duplicate rows
data2_no_duplicates = data2.drop_duplicates()
print(data2_no_duplicates)
print(data2_no_duplicates.duplicated().sum())

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
data1_duplicates = data1.duplicated().any()
print("Are there any duplicate rows in data1? ", data1_duplicates)
print("What are the total number of duplicate rows in data2? ", data2.duplicated().sum())
data1_no_duplicates = data1.drop_duplicates()
print(data1_no_duplicates)
print(data1_no_duplicates.duplicated().sum())


unique_id      int64
al           float64
su           float64
rbc           object
pc            object
pcc           object
ba            object
htn           object
dm            object
cad           object
appet         object
pe            object
ane           object
Target        object
dtype: object
unique_id      int64
age          float64
bp           float64
bgr          float64
bu           float64
sc           float64
sod          float64
pot          float64
hemo         float64
pcv          float64
wbcc         float64
rbcc         float64
dtype: object
Are there any duplicate rows in data2?  True
What are the total number of duplicate rows in data2?  30
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No. If an ID exists in one dataset but not in the other, the combined table might show the information from the present dataset, with the columns from the absent dataset filled with null or missing values (e.g., `NaN` in pandas).

In [49]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
df = pd.merge(data1, data2, on="unique_id", how="outer")
# Print the combined dataset


print(df)


     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
1       103570  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
2       104675  2.0  1.0  normal    normal     present  notpresent   no  yes   
3       105923  1.0  0.0     NaN    normal  notpresent  notpresent  yes  yes   
4       107297  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
487     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
488     996076  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
489     996412  NaN  NaN     NaN       NaN  notpresent  notpresent   no   no   
490     997629  0.0  0.0  normal    normal         NaN         NaN   no   no   
491     997876  0.0  0.0     NaN    normal  notpresent  notpresent   no   no   

     cad  ...    bp    bgr     bu    sc

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [50]:
# Calculate the percentage of rows that contain atleast one missing value
missing_values = df.isnull().sum()
percentage_missing = (missing_values / df.shape[0]) * 100
# Print %
print(percentage_missing)
# Drop these rows from the dataset
df = df.dropna()
# Print the Dataset
print("percentage of missing values in cleaned dataset = ", df.isnull().sum() / df.shape[0] * 100)
print(df)

unique_id     0.000000
al           11.788618
su           13.008130
rbc          38.008130
pc           16.260163
pcc           0.813008
ba            0.813008
htn           0.406504
dm            0.406504
cad           0.406504
appet         0.203252
pe            0.203252
ane           0.203252
Target        0.000000
age           3.658537
bp            6.300813
bgr          11.788618
bu            5.894309
sc            5.487805
sod          21.747967
pot          21.951220
hemo         15.447154
pcv          21.341463
wbcc         29.878049
rbcc         35.569106
dtype: float64
percentage of missing values in cleaned dataset =  unique_id    0.0
al           0.0
su           0.0
rbc          0.0
pc           0.0
pcc          0.0
ba           0.0
htn          0.0
dm           0.0
cad          0.0
appet        0.0
pe           0.0
ane          0.0
Target       0.0
age          0.0
bp           0.0
bgr          0.0
bu           0.0
sc           0.0
sod          0.0
pot          0.0
he

## Part 3.4: Sort the dataset according to the Labels

In [51]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
df = df.sort_values(by="Target").reset_index(drop=True)
# Print the dataset
print(df)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
1       234299  4.0  0.0    normal  abnormal  notpresent  notpresent  yes   
2       636769  2.0  0.0    normal  abnormal     present     present   no   
3       484175  2.0  0.0  abnormal  abnormal     present  notpresent  yes   
4       481293  3.0  1.0    normal  abnormal     present  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
193     433613  0.0  0.0    normal    normal  notpresent  notpresent   no   
194     436056  0.0  0.0    normal    normal  notpresent  notpresent   no   
195     441170  0.0  0.0    normal    normal  notpresent  notpresent   no   
196     441170  0.0  0.0    normal    normal  notpresent  notpresent   no   
197     996076  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm cad  ...    bp    bgr     bu    sc    sod  pot  hemo   pcv     wbc

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [52]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
df_encoded = pd.get_dummies(df, columns=categorical_columns, prefix=categorical_columns)

# Print the dataset
print(df_encoded.head())

   unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
0     102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
1     234299  4.0  0.0  48.0  80.0   95.0  163.0   7.7  136.0  3.8  ...   
2     636769  2.0  0.0  50.0  90.0  128.0  208.0   9.2  134.0  4.8  ...   
3     484175  2.0  0.0  53.0  90.0   70.0  107.0   7.2  114.0  3.7  ...   
4     481293  3.0  1.0  60.0  60.0  288.0   36.0   1.7  130.0  3.0  ...   

   cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
0    True    False       False        True  False    True   False     True   
1    True    False        True       False   True   False   False     True   
2    True    False       False        True  False    True   False     True   
3    True    False       False        True   True   False   False     True   
4    True    False       False        True   True   False   False     True   

   Target_ckd  Target_notckd  
0        True          False  
1        True     

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: If you had a categorical feature like "type of condiment" with five unique values (e.g., ketchup, mustard, mayo, hot sauce, and relish), encoding it with dummy variables would result in five separate columns with the column set to 1 if it is that condiment otherwise 0.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [53]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
def remove_outliers(df, columns):
    for column in columns:
        mean = df[column].mean()
        std_dev = df[column].std()
        # Create a condition for the outliers
        outlier_condition = (df[column] < (mean - 3 * std_dev)) | (df[column] > (mean + 3 * std_dev))
        # Filter the DataFrame
        df = df[~outlier_condition]  # Keep only rows that are not outliers
    return df

df_cleaned = remove_outliers(df_encoded, numerical_columns)


# Print the dataset
print(df_cleaned.head())

    unique_id   al   su   age     bp    bgr     bu   sc    sod  pot  ...  \
7      843362  4.0  0.0  71.0   60.0  118.0  125.0  5.3  136.0  4.9  ...   
8      828592  3.0  0.0  71.0   70.0  219.0   82.0  3.6  133.0  4.4  ...   
17     438182  2.0  0.0  73.0   80.0  253.0  142.0  4.6  138.0  5.8  ...   
22     397388  2.0  2.0  63.0  100.0  280.0   35.0  3.2  143.0  3.5  ...   
25     995177  4.0  3.0  69.0   70.0  214.0   96.0  6.3  120.0  3.9  ...   

    cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
7     True    False       False        True  False    True    True    False   
8    False     True        True       False   True   False    True    False   
17   False     True        True       False   True   False    True    False   
22   False     True        True       False   True   False    True    False   
25   False     True        True       False  False    True   False     True   

    Target_ckd  Target_notckd  
7         True          False  
8   

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [54]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
def normalize_columns(df, columns):
    for column in columns:
        min_value = df[column].min()
        max_value = df[column].max()
        df[column] = (df[column] - min_value) / (max_value - min_value)
    return df
df_normalized = normalize_columns(df_cleaned.copy(), numerical_columns)

# Print the dataset
print(df_normalized.head())

    unique_id   al   su       age    bp       bgr        bu        sc  \
7      843362  4.0  0.0  0.867647  0.00  0.206009  0.871212  0.777778   
8      828592  3.0  0.0  0.867647  0.25  0.639485  0.545455  0.507937   
17     438182  2.0  0.0  0.897059  0.50  0.785408  1.000000  0.666667   
22     397388  2.0  2.0  0.750000  1.00  0.901288  0.189394  0.444444   
25     995177  4.0  3.0  0.838235  0.25  0.618026  0.651515  0.936508   

         sod       pot  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
7   0.533333  0.689655  ...    True    False       False        True  False   
8   0.433333  0.517241  ...   False     True        True       False   True   
17  0.600000  1.000000  ...   False     True        True       False   True   
22  0.766667  0.206897  ...   False     True        True       False   True   
25  0.000000  0.344828  ...   False     True        True       False  False   

    pe_yes  ane_no  ane_yes  Target_ckd  Target_notckd  
7     True    True    False  

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Columns that uniquely identify rows, such as IDs, typically do not contribute to the predictive power of the model. These could be columns like "Patient_ID", "Record_ID", etc., because they do not contain meaningful information about the target variable.

In [55]:
#Remove that column
columns_to_drop = ['unique_id']
df_cleaned = df_normalized.drop(columns=columns_to_drop)

# Print the dataset
print(df_cleaned.columns)
print(df_cleaned.head(15).to_string())  


Index(['al', 'su', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv',
       'wbcc', 'rbcc', 'rbc_abnormal', 'rbc_normal', 'pc_abnormal',
       'pc_normal', 'pcc_notpresent', 'pcc_present', 'ba_notpresent',
       'ba_present', 'htn_no', 'htn_yes', 'dm_no', 'dm_yes', 'cad_no',
       'cad_yes', 'appet_good', 'appet_poor', 'pe_no', 'pe_yes', 'ane_no',
       'ane_yes', 'Target_ckd', 'Target_notckd'],
      dtype='object')
     al   su       age    bp       bgr        bu        sc       sod       pot      hemo       pcv      wbcc      rbcc  rbc_abnormal  rbc_normal  pc_abnormal  pc_normal  pcc_notpresent  pcc_present  ba_notpresent  ba_present  htn_no  htn_yes  dm_no  dm_yes  cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  Target_ckd  Target_notckd
7   4.0  0.0  0.867647  0.00  0.206009  0.871212  0.777778  0.533333  0.689655  0.264368  0.269231  1.000000  0.333333         False        True        False       True            True        False          

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [56]:
# Export the dataframe to a new csv file
df_cleaned.to_csv('cleaned_data.csv', index=False)

# Export the dataframe to a new json file
df_cleaned.to_json('cleaned_data.json', orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

| al  | su   | age     | bp   | bgr     | bu       | sc       | sod     | pot     | hemo    | pcv     | wbcc    | rbcc     | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-----|------|---------|------|---------|----------|----------|---------|---------|---------|---------|---------|----------|--------------|------------|--------------|-----------|----------------|-------------|---------------|------------|--------|---------|-------|--------|--------|---------|------------|------------|-------|--------|--------|---------|------------|----------------|
| 7   | 4.0  | 0.867647 | 0.00 | 0.206009 | 0.871212 | 0.777778 | 0.533333 | 0.689655 | 0.264368 | 0.269231 | 1.000000 | 0.333333 | False        | True       | False        | True      | True           | False       | True          | False      | False  | True    | False | True   | True   | False   | False      | True       | False | True   | True   | False   | True       | False          |
| 8   | 3.0  | 0.867647 | 0.25 | 0.639485 | 0.545455 | 0.507937 | 0.433333 | 0.517241 | 0.149425 | 0.192308 | 0.119266 | 0.121212 | False        | True       | True         | False     | False          | True        | False         | True       | False  | True    | False | True   | False  | True    | True       | False      | True  | False  | True   | False   | True       | False          |
| 17  | 2.0  | 0.897059 | 0.50 | 0.785408 | 1.000000 | 0.666667 | 0.600000 | 1.000000 | 0.160920 | 0.192308 | 0.266055 | 0.333333 | True         | False      | True         | False     | True           | False       | True          | False      | False  | True    | False | True   | True   | False   | True       | False      | True  | False  | True   | False   | True       | False          |
| 22  | 2.0  | 0.750000 | 1.00 | 0.901288 | 0.189394 | 0.444444 | 0.766667 | 0.206897 | 0.448276 | 0.461538 | 0.504587 | 0.303030 | False        | True       | False        | True      | True           | False       | False         | True       | False  | True    | True  | False  | False  | True    | True       | False      | True  | False  | True   | False   | True       | False          |
| 25  | 4.0  | 0.838235 | 0.25 | 0.618026 | 0.651515 | 0.936508 | 0.000000 | 0.344828 | 0.034483 | 0.000000 | 0.660550 | 0.030303 | False        | True       | True         | False     | False          | True        | False         | True       | False  | True    | False | True   | False  | True    | True       | False      | False | True   | False  | True    | True       | False          |
| 27  | 3.0  | 0.897059 | 1.00 | 0.965665 | 0.606061 | 0.825397 | 0.666667 | 0.000000 | 0.011494 | 0.076923 | 0.247706 | 0.000000 | True         | False      | True         | False     | False          | True        | True          | False      | False  | True    | False | True   | False  | True    | False      | True       | True  | False  | True   | False   | True       | False          |
| 31  | 1.0  | 0.500000 | 0.00 | 0.399142 | 0.621212 | 0.460317 | 0.700000 | 0.379310 | 0.080460 | 0.000000 | 0.944954 | 0.000000 | False        | True       | False        | True      | True           | False       | True          | False      | False  | True    | False | True   | True   | False   | True       | False      | True  | False  | True   | False   | True       | False          |
| 32  | 1.0  | 0.691176 | 0.50 | 1.000000 | 0.189394 | 0.142857 | 0.066667 | 0.206897 | 0.149425 | 0.269231 | 0.605505 | 0.333333 | True         | False      | False        | True      | True           | False       | True          | False      | True   | False   | False | True   | True   | False   | False      | True       | True  | False  | True   | False   | True       | False          |
| 35  | 3.0  | 0.632353 | 0.50 | 0.618026 | 0.477273 | 0.555556 | 0.566667 | 0.689655 | 0.206897 | 0.230769 | 0.284404 | 0.151515 | False        | True       | True         | False     | False          | True        | False         | True       | False  | True    | False | True   | True   | False   | True       | False      | False | True   | True   | False   | True       | False          |
| 37  | 3.0  | 0.735294 | 0.25 | 0.223176 | 0.242424 | 0.206349 | 0.533333 | 0.620690 | 0.402299 | 0.423077 | 0.330275 | 0.212121 | False        | True       | True         | False     | True           | False       | True          | False      | False  | True    | False | True   | True   | False   | True       | False      | True  | False  | True   | False   | True       | False          |
| 39  | 2.0  | 0.705882 | 0.75 | 0.150215 | 0.325758 | 0.301587 | 0.533333 | 0.793103 | 0.229885 | 0.192308 | 0.568807 | 0.272727 | True         | False      | False        | True      | True           | False       | True          | False      | True   | False   | True  | False  | True   | False   | True       | False      | True  | False  | True   | False   | True       | False          |
| 40  | 4.0  | 0.647059 | 0.25 | 0.600858 | 0.121212 | 0.206349 | 0.533333 | 0.310345 | 0.804598 | 0.923077 | 0.752294 | 0.727273 | True         | False      | False        | True      | True           | False       | True          | False      | True   | False   | True  | False  | True   | False   | True       | False      | True  | False  | True   | False   | True       | False          |
| 41  | 4.0  | 0.764706 | 0.00 | 0.725322 | 0.363636 | 0.619048 | 0.566667 | 0.862069 | 0.045977 | 0.038462 | 0.293578 | 0.060606 | True         | False      | True         | False     | True           | False       | False         | True       | False  | True    | False | True   | True   | False   | False      | True       | False | True   | True   | False   | True       | False          |
| 42  | 2.0  | 0.647059 | 0.75 | 0.253219 | 0.734848 | 1.000000 | 0.366667 | 0.655172 | 0.000000 | 0.038462 | 0.192661 | 0.060606 | True         | False      | True         | False     | True           | False       | True          | False      | True   | False   | True  | False  | True   | False   | True       | False      | True  | False  | True   | False   | True       | False          |
| 46  | 3.0  | 0.838235 | 0.25 | 0.832618 | 0.583333 | 0.365079 | 0.333333 | 0.379310 | 0.390805 | 0.346154 | 0.486239 | 0.272727 | False        | True       | True         | False     | True           | False       | True          | False      | False  | True    | False | True   | True   | False   | True       | False      | False | True   | True   | False   | True       | False          |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [57]:
# Code to rename all the columns in the dataset
column_mapping = {
    'age': 'age',
    'bp': 'blood_pressure',
    'sg': 'specific_gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc_abnormal': 'red_blood_cells_abnormal',
    'rbc_normal': 'red_blood_cells_normal',
    'pc_abnormal': 'pus_cell_abnormal',
    'pc_normal': 'pus_cell_normal',
    'pcc_notpresent': 'pus_cell_clumps_not_present',
    'pcc_present': 'pus_cell_clumps_present',
    'ba_notpresent': 'bacteria_not_present',
    'ba_present': 'bacteria_present',
    'bgr': 'blood_glucose_random',
    'bu': 'blood_urea',
    'sc': 'serum_creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed_cell_volume',
    'wbcc': 'white_blood_cell_count',
    'rbcc': 'red_blood_cell_count',
    'htn_no': 'hypertension_no',
    'htn_yes': 'hypertension_yes',
    'dm_no': 'diabetes_mellitus_no',
    'dm_yes': 'diabetes_mellitus_yes',
    'cad_no': 'coronary_artery_disease_no',
    'cad_yes': 'coronary_artery_disease_yes',
    'appet_good': 'appetite_good',
    'appet_poor': 'appetite_poor',
    'pe_no': 'pedal_edema_no',
    'pe_yes': 'pedal_edema_yes',
    'ane_no': 'anemia_no',
    'ane_yes': 'anemia_yes',
    'Target_ckd': 'class_ckd',
    'Target_notckd': 'class_not_ckd',
}

# Rename the columns
df_renamed = df_cleaned.rename(columns=column_mapping)

# Print the new column names to verify the changes
print(df_renamed.columns)

Index(['albumin', 'sugar', 'age', 'blood_pressure', 'blood_glucose_random',
       'blood_urea', 'serum_creatinine', 'sodium', 'potassium', 'hemoglobin',
       'packed_cell_volume', 'white_blood_cell_count', 'red_blood_cell_count',
       'red_blood_cells_abnormal', 'red_blood_cells_normal',
       'pus_cell_abnormal', 'pus_cell_normal', 'pus_cell_clumps_not_present',
       'pus_cell_clumps_present', 'bacteria_not_present', 'bacteria_present',
       'hypertension_no', 'hypertension_yes', 'diabetes_mellitus_no',
       'diabetes_mellitus_yes', 'coronary_artery_disease_no',
       'coronary_artery_disease_yes', 'appetite_good', 'appetite_poor',
       'pedal_edema_no', 'pedal_edema_yes', 'anemia_no', 'anemia_yes',
       'class_ckd', 'class_not_ckd'],
      dtype='object')


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [58]:
# Converted SQL to Pandas code
result = df.groupby('Target').size().reset_index(name='count')
print(result)

   Target  count
0     ckd     48
1  notckd    150


This groups the DataFrame by the values in the `Target` column. When you call `size()`, it counts the number of occurrences (rows) for each group. If you use `count()` with the column specified, it counts non-null entries for that column within each group. `reset_index(name='count')`: This converts the resulting Series or DataFrame back into a DataFrame with a new index and renames the counts column to `'count'`.
The result tells us that there are 48 rows with data that indicate chronic kidney disease and 150 that indicate no chronic kidney disease.
